File: messAI/Models/Translation.swift
import Foundation

/// Supported languages for translation
enum LanguageCode: String, CaseIterable, Codable {
    case english = "en"
    case spanish = "es"
    case french = "fr"
    case german = "de"
    case italian = "it"
    case portuguese = "pt"
    case russian = "ru"
    case japanese = "ja"
    case korean = "ko"
    case chinese = "zh"
    case arabic = "ar"
    case hindi = "hi"
    case dutch = "nl"
    case swedish = "sv"
    case norwegian = "no"
    case danish = "da"
    case finnish = "fi"
    case polish = "pl"
    
    /// Human-readable language name
    var displayName: String {
        switch self {
        case .english: return "English"
        case .spanish: return "Spanish"
        case .french: return "French"
        case .german: return "German"
        case .italian: return "Italian"
        case .portuguese: return "Portuguese"
        case .russian: return "Russian"
        case .japanese: return "Japanese"
        case .korean: return "Korean"
        case .chinese: return "Chinese (Simplified)"
        case .arabic: return "Arabic"
        case .hindi: return "Hindi"
        case .dutch: return "Dutch"
        case .swedish: return "Swedish"
        case .norwegian: return "Norwegian"
        case .danish: return "Danish"
        case .finnish: return "Finnish"
        case .polish: return "Polish"
        }
    }
    
    /// Language flag emoji (approximate)
    var flagEmoji: String {
        switch self {
        case .english: return "🇺🇸"
        case .spanish: return "🇪🇸"
        case .french: return "🇫🇷"
        case .german: return "🇩🇪"
        case .italian: return "🇮🇹"
        case .portuguese: return "🇵🇹"
        case .russian: return "🇷🇺"
        case .japanese: return "🇯🇵"
        case .korean: return "🇰🇷"
        case .chinese: return "🇨🇳"
        case .arabic: return "🇸🇦"
        case .hindi: return "🇮🇳"
        case .dutch: return "🇳🇱"
        case .swedish: return "🇸🇪"
        case .norwegian: return "🇳🇴"
        case .danish: return "🇩🇰"
        case .finnish: return "🇫🇮"
        case .polish: return "🇵🇱"
        }
    }
    
    /// Most commonly used languages (for quick access)
    static let popularLanguages: [LanguageCode] = [
        .english, .spanish, .french, .german, .chinese, .japanese
    ]
}

/// Translation method used
enum TranslationMethod: String, Codable {
    case gpt4 = "gpt-4"
    case cached = "cached"
    
    var displayName: String {
        switch self {
        case .gpt4: return "AI Translation"
        case .cached: return "Cached"
        }
    }
}

/// Translation result from AI service
struct TranslationResult: Codable, Identifiable {
    let id = UUID()
    let translatedText: String
    let detectedSourceLanguage: LanguageCode
    let targetLanguage: LanguageCode
    let confidence: Double
    let translationMethod: TranslationMethod
    let processingTimeMs: Int
    let tokensUsed: Int
    let cost: Double
    let timestamp: Date
    
    /// Initialize from Cloud Function response
    init?(from dictionary: [String: Any]) {
        guard let translatedText = dictionary["translatedText"] as? String,
              let detectedSourceLanguageString = dictionary["detectedSourceLanguage"] as? String,
              let detectedSourceLanguage = LanguageCode(rawValue: detectedSourceLanguageString),
              let targetLanguageString = dictionary["targetLanguage"] as? String,
              let targetLanguage = LanguageCode(rawValue: targetLanguageString),
              let confidence = dictionary["confidence"] as? Double,
              let translationMethodString = dictionary["translationMethod"] as? String,
              let translationMethod = TranslationMethod(rawValue: translationMethodString),
              let processingTimeMs = dictionary["processingTimeMs"] as? Int,
              let tokensUsed = dictionary["tokensUsed"] as? Int,
              let cost = dictionary["cost"] as? Double else {
            print("❌ TranslationResult: Failed to parse from dictionary")
            print("   Dictionary: \(dictionary)")
            return nil
        }
        
        self.translatedText = translatedText
        self.detectedSourceLanguage = detectedSourceLanguage
        self.targetLanguage = targetLanguage
        self.confidence = confidence
        self.translationMethod = translationMethod
        self.processingTimeMs = processingTimeMs
        self.tokensUsed = tokensUsed
        self.cost = cost
        self.timestamp = Date()
    }
    
    /// Confidence level description
    var confidenceDescription: String {
        switch confidence {
        case 0.9...1.0: return "Very High"
        case 0.7..<0.9: return "High"
        case 0.5..<0.7: return "Medium"
        case 0.3..<0.5: return "Low"
        default: return "Very Low"
        }
    }
    
    /// Processing time description
    var processingTimeDescription: String {
        if processingTimeMs < 1000 {
            return "\(processingTimeMs)ms"
        } else {
            let seconds = Double(processingTimeMs) / 1000.0
            return String(format: "%.1fs", seconds)
        }
    }
    
    /// Cost description in cents
    var costDescription: String {
        let cents = cost * 100
        if cents < 0.01 {
            return "<$0.01"
        } else {
            return String(format: "$%.3f", cost)
        }
    }
}

/// Translation request data
struct TranslationRequest: Codable {
    let messageText: String
    let targetLanguage: LanguageCode
    let sourceLanguage: LanguageCode?
    let conversationId: String
    let messageId: String
    let preserveFormatting: Bool
    
    init(
        messageText: String,
        targetLanguage: LanguageCode,
        sourceLanguage: LanguageCode? = nil,
        conversationId: String,
        messageId: String,
        preserveFormatting: Bool = true
    ) {
        self.messageText = messageText
        self.targetLanguage = targetLanguage
        self.sourceLanguage = sourceLanguage
        self.conversationId = conversationId
        self.messageId = messageId
        self.preserveFormatting = preserveFormatting
    }
}

/// Translation state for UI
enum TranslationState {
    case idle
    case translating
    case success(TranslationResult)
    case error(String)
    
    var isLoading: Bool {
        if case .translating = self {
            return true
        }
        return false
    }
    
    var translationResult: TranslationResult? {
        if case .success(let result) = self {
            return result
        }
        return nil
    }
    
    var errorMessage: String? {
        if case .error(let message) = self {
            return message
        }
        return nil
    }
}

/// User's preferred languages for translation
struct TranslationPreferences: Codable {
    var defaultTargetLanguage: LanguageCode
    var recentLanguages: [LanguageCode]
    var autoDetectSource: Bool
    var preserveFormatting: Bool
    var showConfidenceScores: Bool
    
    static let `default` = TranslationPreferences(
        defaultTargetLanguage: .english,
        recentLanguages: [],
        autoDetectSource: true,
        preserveFormatting: true,
        showConfidenceScores: false
    )
    
    /// Add a language to recent list (max 5)
    mutating func addRecentLanguage(_ language: LanguageCode) {
        // Remove if already exists
        recentLanguages.removeAll { $0 == language }
        
        // Add to front
        recentLanguages.insert(language, at: 0)
        
        // Keep only 5 most recent
        if recentLanguages.count > 5 {
            recentLanguages = Array(recentLanguages.prefix(5))
        }
    }
}

/// Cache for translations to avoid duplicate API calls
class TranslationCache {
    private var cache: [String: TranslationResult] = [:]
    private let maxCacheSize = 100
    private let cacheExpirationTime: TimeInterval = 24 * 60 * 60 // 24 hours
    
    /// Generate cache key
    private func cacheKey(text: String, targetLanguage: LanguageCode) -> String {
        return "\(text.prefix(100).lowercased())_\(targetLanguage.rawValue)"
    }
    
    /// Get cached translation if available and not expired
    func getTranslation(for text: String, targetLanguage: LanguageCode) -> TranslationResult? {
        let key = cacheKey(text: text, targetLanguage: targetLanguage)
        
        guard let cachedResult = cache[key] else {
            return nil
        }
        
        // Check if expired
        if Date().timeIntervalSince(cachedResult.timestamp) > cacheExpirationTime {
            cache.removeValue(forKey: key)
            return nil
        }
        
        return cachedResult
    }
    
    /// Cache a translation result
    func cacheTranslation(_ result: TranslationResult, for text: String) {
        let key = cacheKey(text: text, targetLanguage: result.targetLanguage)
        cache[key] = result
        
        // Clean up if cache gets too large
        if cache.count > maxCacheSize {
            cleanupCache()
        }
    }
    
    /// Remove expired entries and oldest entries if over limit
    private func cleanupCache() {
        let now = Date()
        
        // Remove expired entries
        cache = cache.filter { _, result in
            now.timeIntervalSince(result.timestamp) <= cacheExpirationTime
        }
        
        // If still over limit, remove oldest entries
        if cache.count > maxCacheSize {
            let sortedByDate = cache.sorted { $0.value.timestamp < $1.value.timestamp }
            let toKeep = sortedByDate.suffix(maxCacheSize - 10) // Keep 10 less than max
            
            cache = Dictionary(uniqueKeysWithValues: toKeep)
        }
    }
    
    /// Clear all cached translations
    func clearCache() {
        cache.removeAll()
    }
    
    /// Get cache statistics
    var cacheInfo: (count: Int, oldestEntry: Date?) {
        let oldestEntry = cache.values.min { $0.timestamp < $1.timestamp }?.timestamp
        return (cache.count, oldestEntry)
    }
}

File: functions/src/ai/translation.ts
import * as functions from 'firebase-functions';
import { OpenAI } from 'openai';

// Initialize OpenAI (get API key from Firebase config)
const openai = new OpenAI({
  apiKey: functions.config().openai.key,
});

/**
 * Supported languages for translation
 */
export const SUPPORTED_LANGUAGES = {
  'en': 'English',
  'es': 'Spanish', 
  'fr': 'French',
  'de': 'German',
  'it': 'Italian',
  'pt': 'Portuguese',
  'ru': 'Russian',
  'ja': 'Japanese',
  'ko': 'Korean',
  'zh': 'Chinese (Simplified)',
  'ar': 'Arabic',
  'hi': 'Hindi',
  'nl': 'Dutch',
  'sv': 'Swedish',
  'no': 'Norwegian',
  'da': 'Danish',
  'fi': 'Finnish',
  'pl': 'Polish'
} as const;

export type LanguageCode = keyof typeof SUPPORTED_LANGUAGES;

/**
 * Translation request data structure
 */
interface TranslationRequest {
  messageText: string;
  targetLanguage: LanguageCode;
  sourceLanguage?: LanguageCode; // Optional, will auto-detect if not provided
  conversationId: string;
  messageId: string;
  preserveFormatting?: boolean;
}

/**
 * Translation response structure
 */
interface TranslationResponse {
  translatedText: string;
  detectedSourceLanguage: LanguageCode;
  targetLanguage: LanguageCode;
  confidence: number;
  translationMethod: 'gpt-4' | 'cached';
  processingTimeMs: number;
  tokensUsed: number;
  cost: number;
}

/**
 * Cache for translation results (in-memory, per function instance)
 * Key: hash of (sourceText + targetLanguage)
 * Value: translation result with timestamp
 */
const translationCache = new Map<string, {
  result: TranslationResponse;
  timestamp: number;
}>();

const CACHE_DURATION_MS = 24 * 60 * 60 * 1000; // 24 hours

/**
 * Generate cache key for translation
 */
function getCacheKey(text: string, targetLang: LanguageCode): string {
  // Simple hash for caching (first 100 chars + target language)
  return `${text.substring(0, 100).toLowerCase()}_${targetLang}`;
}

/**
 * Detect language of given text using GPT-4
 */
async function detectLanguage(text: string): Promise<{
  language: LanguageCode;
  confidence: number;
  tokensUsed: number;
}> {
  const startTime = Date.now();
  
  try {
    const completion = await openai.chat.completions.create({
      model: 'gpt-4',
      messages: [
        {
          role: 'system',
          content: `You are a language detection expert. Analyze the given text and return the language code.
          
Supported languages: ${Object.entries(SUPPORTED_LANGUAGES).map(([code, name]) => `${code} (${name})`).join(', ')}

Respond with ONLY a JSON object in this exact format:
{
  "language": "en",
  "confidence": 0.95
}

Where confidence is between 0.0 and 1.0 representing your certainty.`
        },
        {
          role: 'user',
          content: `Detect the language of this text: "${text}"`
        }
      ],
      max_tokens: 50,
      temperature: 0.1
    });

    const response = completion.choices[0].message.content?.trim();
    if (!response) {
      throw new Error('No response from GPT-4 language detection');
    }

    try {
      const parsed = JSON.parse(response);
      const languageCode = parsed.language as LanguageCode;
      const confidence = Math.min(Math.max(parsed.confidence || 0.5, 0.0), 1.0);

      // Validate language code
      if (!SUPPORTED_LANGUAGES[languageCode]) {
        functions.logger.warn('Unsupported language detected, defaulting to English', {
          detectedLanguage: languageCode,
          originalText: text.substring(0, 50)
        });
        return { language: 'en', confidence: 0.5, tokensUsed: completion.usage?.total_tokens || 0 };
      }

      functions.logger.info('Language detected successfully', {
        language: languageCode,
        confidence,
        processingTimeMs: Date.now() - startTime
      });

      return {
        language: languageCode,
        confidence,
        tokensUsed: completion.usage?.total_tokens || 0
      };

    } catch (parseError) {
      functions.logger.error('Failed to parse language detection response', {
        response,
        error: parseError
      });
      return { language: 'en', confidence: 0.5, tokensUsed: completion.usage?.total_tokens || 0 };
    }

  } catch (error: any) {
    functions.logger.error('Language detection failed', {
      error: error.message,
      text: text.substring(0, 50)
    });
    return { language: 'en', confidence: 0.5, tokensUsed: 0 };
  }
}

/**
 * Translate text using GPT-4
 */
async function translateWithGPT4(
  text: string,
  sourceLanguage: LanguageCode,
  targetLanguage: LanguageCode,
  preserveFormatting: boolean = true
): Promise<{
  translatedText: string;
  tokensUsed: number;
}> {
  const startTime = Date.now();
  
  try {
    const sourceLangName = SUPPORTED_LANGUAGES[sourceLanguage];
    const targetLangName = SUPPORTED_LANGUAGES[targetLanguage];

    const systemPrompt = `You are a professional translator. Translate the given text from ${sourceLangName} to ${targetLangName}.

IMPORTANT RULES:
1. Maintain the original tone and style (casual, formal, emotional, etc.)
2. Preserve any emojis, symbols, or special characters exactly as they are
3. ${preserveFormatting ? 'Preserve line breaks, spacing, and formatting' : 'Output clean, natural text'}
4. If the text contains names, dates, or proper nouns, keep them unchanged unless translation is needed for clarity
5. For messaging context: maintain the conversational, natural feel
6. If the text is already in the target language, return it unchanged
7. Respond with ONLY the translated text, no explanations or meta-commentary

Examples:
- "Hello! 😊 How are you?" → "¡Hola! 😊 ¿Cómo estás?" (Spanish)
- "Meeting at 3pm tomorrow" → "Reunión a las 3pm mañana" (Spanish)
- "Can you pick up Emma from school?" → "¿Puedes recoger a Emma de la escuela?" (Spanish)`;

    const completion = await openai.chat.completions.create({
      model: 'gpt-4',
      messages: [
        {
          role: 'system',
          content: systemPrompt
        },
        {
          role: 'user',
          content: text
        }
      ],
      max_tokens: Math.max(Math.ceil(text.length * 1.5), 100),
      temperature: 0.3 // Low temperature for consistent translations
    });

    const translatedText = completion.choices[0].message.content?.trim();
    if (!translatedText) {
      throw new Error('No translation received from GPT-4');
    }

    functions.logger.info('Translation completed successfully', {
      sourceLanguage,
      targetLanguage,
      originalLength: text.length,
      translatedLength: translatedText.length,
      processingTimeMs: Date.now() - startTime,
      tokensUsed: completion.usage?.total_tokens || 0
    });

    return {
      translatedText,
      tokensUsed: completion.usage?.total_tokens || 0
    };

  } catch (error: any) {
    functions.logger.error('Translation with GPT-4 failed', {
      error: error.message,
      sourceLanguage,
      targetLanguage,
      textLength: text.length
    });
    throw error;
  }
}

/**
 * Main translation function
 * Called by processAI router with 'translation' feature
 */
export async function translateMessage(data: any): Promise<TranslationResponse> {
  const startTime = Date.now();
  
  try {
    // Extract and validate request data
    const {
      messageText,
      targetLanguage,
      sourceLanguage,
      conversationId,
      messageId,
      preserveFormatting = true
    } = data as TranslationRequest;

    // Validate inputs
    if (!messageText || typeof messageText !== 'string') {
      throw new functions.https.HttpsError(
        'invalid-argument',
        'messageText is required and must be a string'
      );
    }

    if (!targetLanguage || !SUPPORTED_LANGUAGES[targetLanguage]) {
      throw new functions.https.HttpsError(
        'invalid-argument',
        `targetLanguage must be one of: ${Object.keys(SUPPORTED_LANGUAGES).join(', ')}`
      );
    }

    if (messageText.trim().length === 0) {
      throw new functions.https.HttpsError(
        'invalid-argument',
        'messageText cannot be empty'
      );
    }

    if (messageText.length > 2000) {
      throw new functions.https.HttpsError(
        'invalid-argument',
        'messageText too long (max 2000 characters)'
      );
    }

    functions.logger.info('Translation request received', {
      conversationId,
      messageId,
      targetLanguage,
      sourceLanguage: sourceLanguage || 'auto-detect',
      textLength: messageText.length,
      preserveFormatting
    });

    // Check cache first
    const cacheKey = getCacheKey(messageText, targetLanguage);
    const cached = translationCache.get(cacheKey);
    
    if (cached && (Date.now() - cached.timestamp) < CACHE_DURATION_MS) {
      functions.logger.info('Translation cache hit', {
        cacheKey: cacheKey.substring(0, 50),
        cacheAge: Date.now() - cached.timestamp
      });
      
      return {
        ...cached.result,
        translationMethod: 'cached',
        processingTimeMs: Date.now() - startTime
      };
    }

    let detectedSourceLanguage: LanguageCode;
    let totalTokensUsed = 0;
    let detectionConfidence = 1.0;

    // Step 1: Detect source language if not provided
    if (!sourceLanguage) {
      const detection = await detectLanguage(messageText);
      detectedSourceLanguage = detection.language;
      detectionConfidence = detection.confidence;
      totalTokensUsed += detection.tokensUsed;

      functions.logger.info('Source language detected', {
        detectedLanguage: detectedSourceLanguage,
        confidence: detectionConfidence
      });
    } else {
      detectedSourceLanguage = sourceLanguage;
    }

    // Step 2: Check if translation is needed
    if (detectedSourceLanguage === targetLanguage) {
      functions.logger.info('No translation needed - same language', {
        language: detectedSourceLanguage
      });
      
      const response: TranslationResponse = {
        translatedText: messageText,
        detectedSourceLanguage,
        targetLanguage,
        confidence: 1.0,
        translationMethod: 'cached',
        processingTimeMs: Date.now() - startTime,
        tokensUsed: totalTokensUsed,
        cost: totalTokensUsed * 0.00003 // Rough GPT-4 cost estimate
      };

      return response;
    }

    // Step 3: Translate with GPT-4
    const translation = await translateWithGPT4(
      messageText,
      detectedSourceLanguage,
      targetLanguage,
      preserveFormatting
    );

    totalTokensUsed += translation.tokensUsed;

    // Step 4: Build response
    const response: TranslationResponse = {
      translatedText: translation.translatedText,
      detectedSourceLanguage,
      targetLanguage,
      confidence: detectionConfidence,
      translationMethod: 'gpt-4',
      processingTimeMs: Date.now() - startTime,
      tokensUsed: totalTokensUsed,
      cost: totalTokensUsed * 0.00003 // Rough GPT-4 cost estimate ($0.03 per 1K tokens)
    };

    // Step 5: Cache the result
    translationCache.set(cacheKey, {
      result: response,
      timestamp: Date.now()
    });

    // Clean up old cache entries (simple cleanup)
    if (translationCache.size > 1000) {
      const now = Date.now();
      for (const [key, value] of translationCache.entries()) {
        if (now - value.timestamp > CACHE_DURATION_MS) {
          translationCache.delete(key);
        }
      }
    }

    functions.logger.info('Translation completed successfully', {
      conversationId,
      sourceLanguage: detectedSourceLanguage,
      targetLanguage,
      confidence: detectionConfidence,
      tokensUsed: totalTokensUsed,
      processingTimeMs: response.processingTimeMs,
      cost: response.cost
    });

    return response;

  } catch (error: any) {
    const processingTime = Date.now() - startTime;
    
    functions.logger.error('Translation failed', {
      error: error.message,
      processingTimeMs: processingTime,
      conversationId: data.conversationId,
      targetLanguage: data.targetLanguage
    });

    // Re-throw HttpsError as-is
    if (error instanceof functions.https.HttpsError) {
      throw error;
    }

    // Handle OpenAI API errors
    if (error.type === 'insufficient_quota') {
      throw new functions.https.HttpsError(
        'resource-exhausted',
        'Translation service temporarily unavailable. Please try again later.'
      );
    }

    if (error.type === 'rate_limit_exceeded') {
      throw new functions.https.HttpsError(
        'resource-exhausted',
        'Too many translation requests. Please try again in a moment.'
      );
    }

    // Generic error
    throw new functions.https.HttpsError(
      'internal',
      'Translation failed. Please try again.'
    );
  }
}

/**
 * Get list of supported languages
 */
export function getSupportedLanguages(): typeof SUPPORTED_LANGUAGES {
  return SUPPORTED_LANGUAGES;
}

/**
 * Clear translation cache (for testing/maintenance)
 */
export function clearTranslationCache(): void {
  translationCache.clear();
  functions.logger.info('Translation cache cleared');
}
